crates/astria-composer/src/composer.rs

```diff
@@ -59,22 +114,42 @@ impl Composer {
         self.api_server.local_addr()
     }
 
+    /// Returns the socker address the grpc collector is served over
+    /// # Error: Returns an error if the listener is not bound
+    pub fn grpc_collector_local_addr(&self) -> io::Result<SocketAddr> {
+        self.grpc_collector_listener.local_addr()
+    }
+
     /// Runs the composer.
     ///
     /// Currently only exits if the api server or searcher stop unexpectedly.
     pub async fn run_until_stopped(self) {
         let Self {
             api_server,
             searcher,
+            grpc_collector_listener,
+            executor_handle,
         } = self;
 
         let api_task =
             tokio::spawn(async move { api_server.await.wrap_err("api server ended unexpectedly") });
         let searcher_task = tokio::spawn(searcher.run());
 
+        // run the grpc server
+        let composer_service = ComposerServiceServer::new(executor_handle);
+        let grpc_server = tonic::transport::Server::builder().add_service(composer_service);
+        let grpc_server_handler = tokio::spawn(async move {
+            grpc_server
+                .serve_with_incoming(tokio_stream::wrappers::TcpListenerStream::new(
+                    grpc_collector_listener,
+                ))
+                .await
```

Comment:
why use the tcp stream instead of just `grpc_server.serve(listen_addr)`?

also, ideally this would use `serve_with_shutdown` (see here https://github.com/astriaorg/astria/blob/main/crates/astria-sequencer/src/sequencer.rs#L199) but it seems composer doesn't have clean shutdown logic yet :(

Reply:
yeah :( that is a future task as of now.

Reply:
also the tcp stream is to be able to support tests which use port 0 to allow the OS to randomly assign a port. We can't use port 0 with just `serve`. TcpStream is mainly being used to enhance testability of the service.